src/app/server/ArlStorage.cpp

```diff
@@ -0,0 +1,158 @@
+using StagingRestriction     = Clusters::AccessControl::Structs::AccessRestrictionStruct::Type;
+
+namespace {
+
+CHIP_ERROR Convert(StagingRestrictionType from, AccessRestrictionProvider::Type & to)
+{
+    switch (from)
+    {
+    case StagingRestrictionType::kAttributeAccessForbidden:
+        to = AccessRestrictionProvider::Type::kAttributeAccessForbidden;
+        break;
+    case StagingRestrictionType::kAttributeWriteForbidden:
+        to = AccessRestrictionProvider::Type::kAttributeWriteForbidden;
+        break;
+    case StagingRestrictionType::kCommandForbidden:
+        to = AccessRestrictionProvider::Type::kCommandForbidden;
+        break;
+    case StagingRestrictionType::kEventForbidden:
+        to = AccessRestrictionProvider::Type::kEventForbidden;
+        break;
+    default:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Convert(AccessRestrictionProvider::Type from, StagingRestrictionType & to)
+{
+    switch (from)
+    {
+    case AccessRestrictionProvider::Type::kAttributeAccessForbidden:
+        to = StagingRestrictionType::kAttributeAccessForbidden;
+        break;
+    case AccessRestrictionProvider::Type::kAttributeWriteForbidden:
+        to = StagingRestrictionType::kAttributeWriteForbidden;
+        break;
+    case AccessRestrictionProvider::Type::kCommandForbidden:
+        to = StagingRestrictionType::kCommandForbidden;
+        break;
+    case AccessRestrictionProvider::Type::kEventForbidden:
+        to = StagingRestrictionType::kEventForbidden;
+        break;
+    default:
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
+} // namespace
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR ArlStorage::DecodableEntry::Decode(TLV::TLVReader & reader)
+{
+    ReturnErrorOnFailure(mStagingEntry.Decode(reader));
+
+    mEntry.fabricIndex    = mStagingEntry.GetFabricIndex();
+    mEntry.endpointNumber = mStagingEntry.endpoint;
+    mEntry.clusterId      = mStagingEntry.cluster;
+
+    auto iterator = mStagingEntry.restrictions.begin();
+    while (iterator.Next())
+    {
+        auto & tmp = iterator.GetValue();
+        AccessRestrictionProvider::Restriction restriction;
+        ReturnErrorOnFailure(Convert(tmp.type, restriction.restrictionType));
+
+        if (!tmp.id.IsNull())
+        {
+            restriction.id.SetValue(tmp.id.Value());
+        }
+
+        mEntry.restrictions.push_back(restriction);
+    }
+    ReturnErrorOnFailure(iterator.GetStatus());
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ArlStorage::EncodableEntry::EncodeForRead(TLV::TLVWriter & writer, TLV::Tag tag, FabricIndex fabric) const
+{
+    ReturnErrorOnFailure(Stage());
+    ReturnErrorOnFailure(mStagingEntry.EncodeForRead(writer, tag, fabric));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ArlStorage::EncodableEntry::EncodeForWrite(TLV::TLVWriter & writer, TLV::Tag tag) const
```

Comment:
So this is used to store the entries in persistent storage, right?

This part is very much not making sense to me.  `mStagingEntry.EncodeForWrite` will not encode a fabric index (it's designed to be used for attribute writes, and per spec those do not include fabric indices on fabric scoped structs).  Then when we read from storage we do a `Decode()`, never set the fabric index on the struct (right?) but then later code assumes those structs have a fabric index set.

This all needs a _lot_ more tests.

Reply:
Fixed.